blockstream.go

```diff
@@ -97,8 +101,15 @@ func (bs *BlockStream) Session(ctx context.Context, token access.Token, autosave
 			defer bs.wg.Done()
 
 			if err := f(); err != nil {
-				log.Error(err)
 				s.Reset()
+				ses.removeProvider()
+
+				log.Error(err)
+				if ses.prvs == 0 {
```

Comment:
Reading providers is non-atomic. Pls move read to the new session method and do it atomically:
```go
atomic.LoadUint32(&ses.prvs)
```